tensorboardX/record_writer.py

```diff
@@ -3,14 +3,102 @@
 The code was borrow from https://github.com/TeamHG-Memex/tensorboard_logger
 """
 
+import io
+import os.path
 import re
 import struct
+try:
+    import boto3
+    S3_ENABLED = True
+except ImportError:
+    S3_ENABLED = False
 
 from .crc32c import crc32c
 
+
 _VALID_OP_NAME_START = re.compile('^[A-Za-z0-9.]')
 _VALID_OP_NAME_PART = re.compile('[A-Za-z0-9_.\\-/]+')
 
+# Registry of writer factories by prefix backends.
+#
+# Currently supports "s3://" URLs for S3 based on boto and falls
+# back to local filesystem.
+REGISTERED_FACTORIES = {}
+
+
+def register_writer_factory(prefix, factory):
+    if ':' in prefix:
+        raise ValueError('prefix cannot contain a :')
+    REGISTERED_FACTORIES[prefix] = factory
+
+
+def directory_check(path):
+    '''Initialize the directory for log files.'''
+    try:
+        prefix = path.split(':')[0]
+        factory = REGISTERED_FACTORIES[prefix]
+        return factory.directory_check(path)
+    except KeyError:
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+
+def open_file(path):
+    '''Open a writer for outputting event files.'''
+    try:
+        prefix = path.split(':')[0]
+        factory = REGISTERED_FACTORIES[prefix]
+        return factory.open(path, 'wb')
+    except KeyError:
+        return open(path, 'wb')
+
+
+class S3RecordWriter(object):
+    """Writes tensorboard protocol buffer files to S3."""
+
+    def __init__(self, path, mode):
+        if not S3_ENABLED:
+            raise ImportError("boto3 must be installed for S3 support.")
+        if mode != "wb" and mode != "bw":
```

Comment:
I have not use boto before. But I guess `mode` be used as indicator while opening the s3 object?

Reply:
Actually the mode we're talking about here is the one passed from `open_file`, but this actually doesn't make sense, since we always pass the same mode. I'll clean this up.